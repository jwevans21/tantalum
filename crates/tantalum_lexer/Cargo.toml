[package]
name = "tantalum_lexer"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}
pretty_assertions = { workspace = true }
rstest = { workspace = true }


[lints]
workspace = true
