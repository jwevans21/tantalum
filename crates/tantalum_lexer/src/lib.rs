//! # Tantalum Lexer
//!
//! This crate provides a lexer for the Tantalum language.

use tantalum_source::{SourceFile, SourceLocation, SourceSpan};
use tantalum_syntax::{SyntaxKind, SyntaxToken};

/// A lexer for the Tantalum language.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Lexer<'file> {
    /// The source file that the lexer is currently lexing.
    file: &'file SourceFile,
    /// The current location in the source file.
    location: SourceLocation,
}

impl<'file> Lexer<'file> {
    /// Create a new lexer from a source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let lexer = Lexer::new(file);
    /// ```
    #[must_use]
    pub fn new(file: &'file SourceFile) -> Self {
        Self {
            file,
            location: file.location_start(),
        }
    }

    /// Get the source file that the lexer is currently lexing.
    #[must_use]
    pub fn source(&self) -> &'file SourceFile {
        self.file
    }

    /// Get the current location in the source file.
    #[must_use]
    pub fn location(&self) -> SourceLocation {
        self.location
    }

    /// Get the character at a location in the source file.
    fn at_location(&self, location: SourceLocation) -> Option<char> {
        self.file.read_char(location)
    }

    /// Extract the next token from the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.next_token().expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    /// ```
    #[must_use]
    pub fn next_token(&mut self) -> Option<SyntaxToken> {
        let start = self.location;

        let (kind, end) = self.characters_to_token(start)?;

        self.location = end;

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Peek at the next token in the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.peek_token(1).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    ///
    /// let token = lexer.peek_token(2).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Whitespace);
    ///
    /// let token = lexer.peek_token(3).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Identifier);
    /// ```
    #[must_use]
    pub fn peek_token(&self, mut n: usize) -> Option<SyntaxToken> {
        let mut start = self.location;

        let (kind, end) = loop {
            n -= 1;
            let (kind, end) = self.characters_to_token(start)?;

            if n == 0 {
                break (kind, end);
            }

            start = end;
        };

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Convert a character to a [`SyntaxKind`].
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    fn characters_to_token(&self, start: SourceLocation) -> Option<(SyntaxKind, SourceLocation)> {
        let mut current = start;

        let c = self.at_location(current)?;
        current = current.next_by(c.len_utf8());

        let token = match c {
            '(' => SyntaxKind::LParen,
            ')' => SyntaxKind::RParen,
            '{' => SyntaxKind::LBrace,
            '}' => SyntaxKind::RBrace,
            '[' => SyntaxKind::LBracket,
            ']' => SyntaxKind::RBracket,
            '<' => SyntaxKind::LAngle,
            '>' => SyntaxKind::RAngle,
            ',' => SyntaxKind::Comma,
            ':' => SyntaxKind::Colon,
            ';' => SyntaxKind::Semicolon,
            '.' => SyntaxKind::Dot,
            '&' => SyntaxKind::Ampersand,
            '|' => SyntaxKind::Pipe,
            '^' => SyntaxKind::Caret,
            '+' => SyntaxKind::Plus,
            '-' => SyntaxKind::Minus,
            '*' => SyntaxKind::Star,
            '/' => SyntaxKind::Slash,
            '\\' => SyntaxKind::Backslash,
            '%' => SyntaxKind::Percent,
            '=' => SyntaxKind::Equals,
            '!' => SyntaxKind::Bang,
            '~' => SyntaxKind::Tilde,
            '?' => SyntaxKind::Question,
            '@' => SyntaxKind::At,
            '0'..='9' => SyntaxKind::Digit,
            '_' | 'a'..='z' | 'A'..='Z' => {
                while let Some(c) = self.at_location(current) {
                    match c {
                        '_' | 'a'..='z' | 'A'..='Z' | '0'..='9' => {
                            current = current.next_by(c.len_utf8());
                        }
                        _ => break,
                    }
                }

                match self.file.read_span(SourceSpan::new(
                    self.file.id(),
                    self.location.offset(),
                    current.offset(),
                ))? {
                    "fn" => SyntaxKind::Fn,
                    "let" => SyntaxKind::Let,
                    "return" => SyntaxKind::Return,
                    "if" => SyntaxKind::If,
                    "else" => SyntaxKind::Else,
                    "true" => SyntaxKind::True,
                    "false" => SyntaxKind::False,
                    _ => SyntaxKind::Identifier,
                }
            }
            c if c.is_whitespace() => SyntaxKind::Whitespace,
            _ => SyntaxKind::Other,
        };

        Some((token, current))
    }
}

#[test]
fn keywords() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "fn let return if else true false".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 7);

    assert_eq!(tokens[0].kind(), SyntaxKind::Fn);
    assert_eq!(tokens[1].kind(), SyntaxKind::Let);
    assert_eq!(tokens[2].kind(), SyntaxKind::Return);
    assert_eq!(tokens[3].kind(), SyntaxKind::If);
    assert_eq!(tokens[4].kind(), SyntaxKind::Else);
    assert_eq!(tokens[5].kind(), SyntaxKind::True);
    assert_eq!(tokens[6].kind(), SyntaxKind::False);
}

#[test]
fn identifiers() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "a A _ a1 A1 _1 _a _A __AA01".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 9);

    assert_eq!(tokens[0].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[1].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[2].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[3].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[4].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[5].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[6].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[7].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[8].kind(), SyntaxKind::Identifier);
}
