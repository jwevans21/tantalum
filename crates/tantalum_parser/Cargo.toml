[package]
name = "tantalum_parser"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_ast = { path = "../tantalum_ast" }
tantalum_lexer = { path = "../tantalum_lexer" }
tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}

[lints]
workspace = true
