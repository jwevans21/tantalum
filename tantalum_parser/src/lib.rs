use error::ParseError;
use tantalum_ast::TopLevelExpression;
use tantalum_lexer::{token::Token, token_kind::TokenKind, Lexer};
use tantalum_span::Span;

pub mod error;

mod expressions;
mod statements;
mod types;
mod top_level;

#[cfg(test)]
mod tests;

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct Parser<'file_name, 'source> {
    // lexer: Lexer<'file_name, 'source>,
    source: &'source str,
    file_name: &'file_name str,
    tokens: Vec<Token<'file_name, 'source>>,
    eof: Span<'file_name>,
    position: usize,
}

impl<'file_name, 'source> Parser<'file_name, 'source> {
    #[must_use]
    #[inline]
    pub fn new(mut lexer: Lexer<'file_name, 'source>) -> Self {
        Self {
            source: lexer.source(),
            file_name: lexer.file_name(),
            tokens: lexer.by_ref().collect(),
            eof: lexer.span(),
            position: 0,
        }
    }

    /// Parse the entire source file.
    /// 
    /// # Errors
    /// 
    /// Returns an error if the parser encounters an unexpected token or the end of the file.
    pub fn parse(
        &mut self,
    ) -> Result<Vec<TopLevelExpression<'file_name, 'source>>, error::ParseError<'file_name, 'source>>
    {
        let mut top_levels = Vec::new();

        while !self.is_eof() {
            top_levels.push(self.parse_top_level()?);
        }

        Ok(top_levels)
    }

    fn is_eof(&self) -> bool {
        self.position >= self.tokens.len()
    }

    fn is_at(&self, kind: TokenKind) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position)
            .filter(|token| token.kind() == kind)
            .copied()
    }

    fn advance_if(&mut self, kind: TokenKind) -> Option<Token<'file_name, 'source>> {
        if let Some(token) = self.is_at(kind) {
            self.position += 1;
            Some(token)
        } else {
            None
        }
    }

    fn expect(
        &mut self,
        kind: TokenKind,
    ) -> Result<Token<'file_name, 'source>, error::ParseError<'file_name, 'source>> {
        let Some(token) = self.tokens.get(self.position) else {
            return Err(error::ParseError::unexpected_eof(self.source, self.eof));
        };

        if token.kind() == kind {
            self.position += 1;
            Ok(*token)
        } else {
            Err(error::ParseError::unexpected_token_set(
                self.source,
                token.span(),
                token.kind(),
                &[kind],
            ))
        }
    }

    #[allow(unused)]
    fn next_is(&self, kind: TokenKind) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position + 1)
            .filter(|token| token.kind() == kind)
            .copied()
    }

    fn is_at_any<'a>(&self, set: &'a [TokenKind]) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position)
            .filter(|token| set.contains(&token.kind()))
            .copied()
    }

    fn advance_if_any<'a>(&mut self, set: &'a [TokenKind]) -> Option<Token<'file_name, 'source>> {
        if let Some(token) = self.is_at_any(set) {
            self.position += 1;
            Some(token)
        } else {
            None
        }
    }

    fn expect_any<'a>(
        &mut self,
        set: &'a [TokenKind],
    ) -> Result<Token<'file_name, 'source>, error::ParseError<'file_name, 'source>> {
        let Some(token) = self.tokens.get(self.position) else {
            return Err(error::ParseError::unexpected_eof(self.source, self.eof));
        };

        if set.contains(&token.kind()) {
            self.position += 1;
            Ok(*token)
        } else {
            Err(error::ParseError::unexpected_token_set(
                self.source,
                token.span(),
                token.kind(),
                set,
            ))
        }
    }

    #[allow(unused)]
    fn next_is_any<'a>(&self, set: &'a [TokenKind]) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position + 1)
            .filter(|token| set.contains(&token.kind()))
            .copied()
    }

    fn nth(&self, n: usize) -> Option<Token<'file_name, 'source>> {
        self.tokens.get(self.position + n).copied()
    }

    fn put_back(&mut self, n: usize) {
        self.position -= n;
    }
}
