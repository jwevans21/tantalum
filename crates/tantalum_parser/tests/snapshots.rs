use insta::assert_ron_snapshot;

use tantalum_lexer::Lexer;
use tantalum_parser::Parser;

macro_rules! snapshot_from_file {
    ($file_name:ident) => {
        #[test]
        fn $file_name() {
            let source = include_str!(concat!("../../examples/", stringify!($file_name), ".ta"));
            let lexer = Lexer::new(concat!(stringify!($file_name), ".ta"), source);

            let mut parser = Parser::new(lexer);

            let ast = parser.parse();

            assert_ron_snapshot!(ast);
        }
    };
    ($($file_name:ident),*) => {
        $(snapshot_from_file! { $file_name })*
    };
}

snapshot_from_file! {
    start,
    hello_world,
    conditionals,
    loops
}
