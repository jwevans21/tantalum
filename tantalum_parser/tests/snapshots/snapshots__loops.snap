---
source: tantalum_parser/tests/snapshots.rs
expression: ast
snapshot_kind: text
---
Ok([
  TopLevelExpression(
    span: Span(
      file_name: "loops.ta",
      start: 0,
      end: 46,
      line: 1,
      column: 1,
    ),
    kind: ExternalFunction(
      name: "printf",
      parameters: [
        Parameter(
          span: Span(
            file_name: "loops.ta",
            start: 17,
            end: 34,
            line: 1,
            column: 18,
          ),
          kind: Named(
            name: "format",
            ty: Type(
              span: Span(
                file_name: "loops.ta",
                start: 25,
                end: 34,
                line: 1,
                column: 26,
              ),
              kind: Pointer(Type(
                span: Span(
                  file_name: "loops.ta",
                  start: 26,
                  end: 34,
                  line: 1,
                  column: 27,
                ),
                kind: Const(Type(
                  span: Span(
                    file_name: "loops.ta",
                    start: 32,
                    end: 34,
                    line: 1,
                    column: 33,
                  ),
                  kind: Named("u8"),
                )),
              )),
            ),
          ),
        ),
      ],
      return_type: Type(
        span: Span(
          file_name: "loops.ta",
          start: 42,
          end: 45,
          line: 1,
          column: 43,
        ),
        kind: Named("i32"),
      ),
      is_variadic: true,
    ),
  ),
  TopLevelExpression(
    span: Span(
      file_name: "loops.ta",
      start: 48,
      end: 397,
      line: 3,
      column: 0,
    ),
    kind: FunctionDeclaration(
      name: "main",
      parameters: [],
      return_type: Type(
        span: Span(
          file_name: "loops.ta",
          start: 59,
          end: 62,
          line: 3,
          column: 11,
        ),
        kind: Named("i32"),
      ),
      body: Statement(
        span: Span(
          file_name: "loops.ta",
          start: 63,
          end: 397,
          line: 3,
          column: 15,
        ),
        kind: Block(
          statements: [
            Statement(
              span: Span(
                file_name: "loops.ta",
                start: 69,
                end: 85,
                line: 4,
                column: 4,
              ),
              kind: VariableDeclaration(
                name: "a",
                ty: Some(Type(
                  span: Span(
                    file_name: "loops.ta",
                    start: 76,
                    end: 79,
                    line: 4,
                    column: 11,
                  ),
                  kind: Named("i32"),
                )),
                value: Expression(
                  span: Span(
                    file_name: "loops.ta",
                    start: 82,
                    end: 84,
                    line: 4,
                    column: 17,
                  ),
                  kind: IntegerLiteral(
                    value: "10",
                    radix: 10,
                  ),
                ),
              ),
            ),
            Statement(
              span: Span(
                file_name: "loops.ta",
                start: 90,
                end: 106,
                line: 5,
                column: 4,
              ),
              kind: VariableDeclaration(
                name: "b",
                ty: Some(Type(
                  span: Span(
                    file_name: "loops.ta",
                    start: 97,
                    end: 100,
                    line: 5,
                    column: 11,
                  ),
                  kind: Named("i32"),
                )),
                value: Expression(
                  span: Span(
                    file_name: "loops.ta",
                    start: 103,
                    end: 105,
                    line: 5,
                    column: 17,
                  ),
                  kind: IntegerLiteral(
                    value: "20",
                    radix: 10,
                  ),
                ),
              ),
            ),
            Statement(
              span: Span(
                file_name: "loops.ta",
                start: 111,
                end: 127,
                line: 6,
                column: 4,
              ),
              kind: VariableDeclaration(
                name: "c",
                ty: Some(Type(
                  span: Span(
                    file_name: "loops.ta",
                    start: 118,
                    end: 121,
                    line: 6,
                    column: 11,
                  ),
                  kind: Named("i32"),
                )),
                value: Expression(
                  span: Span(
                    file_name: "loops.ta",
                    start: 124,
                    end: 126,
                    line: 6,
                    column: 17,
                  ),
                  kind: IntegerLiteral(
                    value: "30",
                    radix: 10,
                  ),
                ),
              ),
            ),
            Statement(
              span: Span(
                file_name: "loops.ta",
                start: 133,
                end: 225,
                line: 8,
                column: 4,
              ),
              kind: While(
                condition: Expression(
                  span: Span(
                    file_name: "loops.ta",
                    start: 139,
                    end: 144,
                    line: 8,
                    column: 10,
                  ),
                  kind: BinaryOperation(
                    left: Expression(
                      span: Span(
                        file_name: "loops.ta",
                        start: 139,
                        end: 140,
                        line: 8,
                        column: 10,
                      ),
                      kind: Variable(
                        name: "b",
                      ),
                    ),
                    operator: BinaryOperator(
                      span: Span(
                        file_name: "loops.ta",
                        start: 141,
                        end: 142,
                        line: 8,
                        column: 12,
                      ),
                      kind: GreaterThan,
                    ),
                    right: Expression(
                      span: Span(
                        file_name: "loops.ta",
                        start: 143,
                        end: 144,
                        line: 8,
                        column: 14,
                      ),
                      kind: Variable(
                        name: "a",
                      ),
                    ),
                  ),
                ),
                body: Statement(
                  span: Span(
                    file_name: "loops.ta",
                    start: 145,
                    end: 225,
                    line: 8,
                    column: 16,
                  ),
                  kind: Block(
                    statements: [
                      Statement(
                        span: Span(
                          file_name: "loops.ta",
                          start: 155,
                          end: 165,
                          line: 9,
                          column: 8,
                        ),
                        kind: Expression(
                          expression: Expression(
                            span: Span(
                              file_name: "loops.ta",
                              start: 155,
                              end: 164,
                              line: 9,
                              column: 8,
                            ),
                            kind: BinaryOperation(
                              left: Expression(
                                span: Span(
                                  file_name: "loops.ta",
                                  start: 155,
                                  end: 156,
                                  line: 9,
                                  column: 8,
                                ),
                                kind: Variable(
                                  name: "a",
                                ),
                              ),
                              operator: BinaryOperator(
                                span: Span(
                                  file_name: "loops.ta",
                                  start: 157,
                                  end: 158,
                                  line: 9,
                                  column: 10,
                                ),
                                kind: Assignment,
                              ),
                              right: Expression(
                                span: Span(
                                  file_name: "loops.ta",
                                  start: 159,
                                  end: 164,
                                  line: 9,
                                  column: 12,
                                ),
                                kind: BinaryOperation(
                                  left: Expression(
                                    span: Span(
                                      file_name: "loops.ta",
                                      start: 159,
                                      end: 160,
                                      line: 9,
                                      column: 12,
                                    ),
                                    kind: Variable(
                                      name: "a",
                                    ),
                                  ),
                                  operator: BinaryOperator(
                                    span: Span(
                                      file_name: "loops.ta",
                                      start: 161,
                                      end: 162,
                                      line: 9,
                                      column: 14,
                                    ),
                                    kind: Addition,
                                  ),
                                  right: Expression(
                                    span: Span(
                                      file_name: "loops.ta",
                                      start: 163,
                                      end: 164,
                                      line: 9,
                                      column: 16,
                                    ),
                                    kind: IntegerLiteral(
                                      value: "1",
                                      radix: 10,
                                    ),
                                  ),
                                ),
                              ),
                            ),
                          ),
                        ),
                      ),
                      Statement(
                        span: Span(
                          file_name: "loops.ta",
                          start: 175,
                          end: 219,
                          line: 11,
                          column: 8,
                        ),
                        kind: If(
                          condition: Expression(
                            span: Span(
                              file_name: "loops.ta",
                              start: 178,
                              end: 185,
                              line: 11,
                              column: 11,
                            ),
                            kind: BinaryOperation(
                              left: Expression(
                                span: Span(
                                  file_name: "loops.ta",
                                  start: 178,
                                  end: 179,
                                  line: 11,
                                  column: 11,
                                ),
                                kind: Variable(
                                  name: "a",
                                ),
                              ),
                              operator: BinaryOperator(
                                span: Span(
                                  file_name: "loops.ta",
                                  start: 180,
                                  end: 182,
                                  line: 11,
                                  column: 13,
                                ),
                                kind: Equality,
                              ),
                              right: Expression(
                                span: Span(
                                  file_name: "loops.ta",
                                  start: 183,
                                  end: 185,
                                  line: 11,
                                  column: 16,
                                ),
                                kind: IntegerLiteral(
                                  value: "15",
                                  radix: 10,
                                ),
                              ),
                            ),
                          ),
                          body: Statement(
                            span: Span(
                              file_name: "loops.ta",
                              start: 186,
                              end: 219,
                              line: 11,
                              column: 19,
                            ),
                            kind: Block(
                              statements: [
                                Statement(
                                  span: Span(
                                    file_name: "loops.ta",
                                    start: 200,
                                    end: 209,
                                    line: 12,
                                    column: 12,
                                  ),
                                  kind: Continue,
                                ),
                              ],
                            ),
                          ),
                          else_branch: None,
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
            Statement(
              span: Span(
                file_name: "loops.ta",
                start: 231,
                end: 252,
                line: 16,
                column: 4,
              ),
              kind: Expression(
                expression: Expression(
                  span: Span(
                    file_name: "loops.ta",
                    start: 231,
                    end: 251,
                    line: 16,
                    column: 4,
                  ),
                  kind: FunctionCall(
                    source: Expression(
                      span: Span(
                        file_name: "loops.ta",
                        start: 231,
                        end: 237,
                        line: 16,
                        column: 4,
                      ),
                      kind: Variable(
                        name: "printf",
                      ),
                    ),
                    arguments: [
                      Expression(
                        span: Span(
                          file_name: "loops.ta",
                          start: 238,
                          end: 247,
                          line: 16,
                          column: 11,
                        ),
                        kind: StringLiteral(
                          value: "\"a: %d\\n\"",
                        ),
                      ),
                      Expression(
                        span: Span(
                          file_name: "loops.ta",
                          start: 249,
                          end: 250,
                          line: 16,
                          column: 22,
                        ),
                        kind: Variable(
                          name: "a",
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
            Statement(
              span: Span(
                file_name: "loops.ta",
                start: 258,
                end: 333,
                line: 18,
                column: 4,
              ),
              kind: ForInitCondUpdate(
                initialization: Statement(
                  span: Span(
                    file_name: "loops.ta",
                    start: 262,
                    end: 277,
                    line: 18,
                    column: 8,
                  ),
                  kind: VariableDeclaration(
                    name: "i",
                    ty: Some(Type(
                      span: Span(
                        file_name: "loops.ta",
                        start: 269,
                        end: 272,
                        line: 18,
                        column: 15,
                      ),
                      kind: Named("i32"),
                    )),
                    value: Expression(
                      span: Span(
                        file_name: "loops.ta",
                        start: 275,
                        end: 276,
                        line: 18,
                        column: 21,
                      ),
                      kind: Variable(
                        name: "a",
                      ),
                    ),
                  ),
                ),
                condition: Statement(
                  span: Span(
                    file_name: "loops.ta",
                    start: 278,
                    end: 284,
                    line: 18,
                    column: 24,
                  ),
                  kind: Expression(
                    expression: Expression(
                      span: Span(
                        file_name: "loops.ta",
                        start: 278,
                        end: 283,
                        line: 18,
                        column: 24,
                      ),
                      kind: BinaryOperation(
                        left: Expression(
                          span: Span(
                            file_name: "loops.ta",
                            start: 278,
                            end: 279,
                            line: 18,
                            column: 24,
                          ),
                          kind: Variable(
                            name: "i",
                          ),
                        ),
                        operator: BinaryOperator(
                          span: Span(
                            file_name: "loops.ta",
                            start: 280,
                            end: 281,
                            line: 18,
                            column: 26,
                          ),
                          kind: LessThan,
                        ),
                        right: Expression(
                          span: Span(
                            file_name: "loops.ta",
                            start: 282,
                            end: 283,
                            line: 18,
                            column: 28,
                          ),
                          kind: Variable(
                            name: "c",
                          ),
                        ),
                      ),
                    ),
                  ),
                ),
                update: Statement(
                  span: Span(
                    file_name: "loops.ta",
                    start: 285,
                    end: 295,
                    line: 18,
                    column: 31,
                  ),
                  kind: Expression(
                    expression: Expression(
                      span: Span(
                        file_name: "loops.ta",
                        start: 285,
                        end: 294,
                        line: 18,
                        column: 31,
                      ),
                      kind: BinaryOperation(
                        left: Expression(
                          span: Span(
                            file_name: "loops.ta",
                            start: 285,
                            end: 286,
                            line: 18,
                            column: 31,
                          ),
                          kind: Variable(
                            name: "i",
                          ),
                        ),
                        operator: BinaryOperator(
                          span: Span(
                            file_name: "loops.ta",
                            start: 287,
                            end: 288,
                            line: 18,
                            column: 33,
                          ),
                          kind: Assignment,
                        ),
                        right: Expression(
                          span: Span(
                            file_name: "loops.ta",
                            start: 289,
                            end: 294,
                            line: 18,
                            column: 35,
                          ),
                          kind: BinaryOperation(
                            left: Expression(
                              span: Span(
                                file_name: "loops.ta",
                                start: 289,
                                end: 290,
                                line: 18,
                                column: 35,
                              ),
                              kind: Variable(
                                name: "i",
                              ),
                            ),
                            operator: BinaryOperator(
                              span: Span(
                                file_name: "loops.ta",
                                start: 291,
                                end: 292,
                                line: 18,
                                column: 37,
                              ),
                              kind: Addition,
                            ),
                            right: Expression(
                              span: Span(
                                file_name: "loops.ta",
                                start: 293,
                                end: 294,
                                line: 18,
                                column: 39,
                              ),
                              kind: IntegerLiteral(
                                value: "1",
                                radix: 10,
                              ),
                            ),
                          ),
                        ),
                      ),
                    ),
                  ),
                ),
                body: Statement(
                  span: Span(
                    file_name: "loops.ta",
                    start: 296,
                    end: 333,
                    line: 18,
                    column: 42,
                  ),
                  kind: Block(
                    statements: [
                      Statement(
                        span: Span(
                          file_name: "loops.ta",
                          start: 306,
                          end: 327,
                          line: 19,
                          column: 8,
                        ),
                        kind: Expression(
                          expression: Expression(
                            span: Span(
                              file_name: "loops.ta",
                              start: 306,
                              end: 326,
                              line: 19,
                              column: 8,
                            ),
                            kind: FunctionCall(
                              source: Expression(
                                span: Span(
                                  file_name: "loops.ta",
                                  start: 306,
                                  end: 312,
                                  line: 19,
                                  column: 8,
                                ),
                                kind: Variable(
                                  name: "printf",
                                ),
                              ),
                              arguments: [
                                Expression(
                                  span: Span(
                                    file_name: "loops.ta",
                                    start: 313,
                                    end: 322,
                                    line: 19,
                                    column: 15,
                                  ),
                                  kind: StringLiteral(
                                    value: "\"i: %d\\n\"",
                                  ),
                                ),
                                Expression(
                                  span: Span(
                                    file_name: "loops.ta",
                                    start: 324,
                                    end: 325,
                                    line: 19,
                                    column: 26,
                                  ),
                                  kind: Variable(
                                    name: "i",
                                  ),
                                ),
                              ],
                            ),
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
            Statement(
              span: Span(
                file_name: "loops.ta",
                start: 339,
                end: 380,
                line: 22,
                column: 4,
              ),
              kind: Expression(
                expression: Expression(
                  span: Span(
                    file_name: "loops.ta",
                    start: 339,
                    end: 379,
                    line: 22,
                    column: 4,
                  ),
                  kind: FunctionCall(
                    source: Expression(
                      span: Span(
                        file_name: "loops.ta",
                        start: 339,
                        end: 345,
                        line: 22,
                        column: 4,
                      ),
                      kind: Variable(
                        name: "printf",
                      ),
                    ),
                    arguments: [
                      Expression(
                        span: Span(
                          file_name: "loops.ta",
                          start: 346,
                          end: 369,
                          line: 22,
                          column: 11,
                        ),
                        kind: StringLiteral(
                          value: "\"a: %d, b: %d, c: %d\\n\"",
                        ),
                      ),
                      Expression(
                        span: Span(
                          file_name: "loops.ta",
                          start: 371,
                          end: 372,
                          line: 22,
                          column: 36,
                        ),
                        kind: Variable(
                          name: "a",
                        ),
                      ),
                      Expression(
                        span: Span(
                          file_name: "loops.ta",
                          start: 374,
                          end: 375,
                          line: 22,
                          column: 39,
                        ),
                        kind: Variable(
                          name: "b",
                        ),
                      ),
                      Expression(
                        span: Span(
                          file_name: "loops.ta",
                          start: 377,
                          end: 378,
                          line: 22,
                          column: 42,
                        ),
                        kind: Variable(
                          name: "c",
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ),
            Statement(
              span: Span(
                file_name: "loops.ta",
                start: 386,
                end: 395,
                line: 24,
                column: 4,
              ),
              kind: Return(
                value: Some(Expression(
                  span: Span(
                    file_name: "loops.ta",
                    start: 393,
                    end: 394,
                    line: 24,
                    column: 11,
                  ),
                  kind: IntegerLiteral(
                    value: "0",
                    radix: 10,
                  ),
                )),
              ),
            ),
          ],
        ),
      ),
    ),
  ),
])
